### DISCUSSION TITLE
tracking down PG errors

### DISCUSSION THREAD

**Discussion by Alex-Jordan** [openwebwork/pg#722] [Ideas]
Sometimes an exercise file has some problem with it that leads to an entry in the apache error log. And sometimes, the message is not very helpful for tracking down the issue. Examples:

```
Use of uninitialized value $expression in concatenation (.) or string at (eval 2138) line 28.
Odd number of elements in anonymous hash at (eval 3493) line 2.
Subroutine install_problem_grader redefined at (eval 3382) line 2, <MACROFILE> chunk 1.
```

Is there anything we can do to improve these entries to help actually identify the offending .pg file?

The last one of these examples comes from a problem using problemRandomize.pl, and it's that macro file that has the issue. At least we see "MACROFILE", but it would be nice to get a report of _which_ macro file.

Related. Sometimes an error message is just natural in the course of editing a problem. Or as was the case with the first example above, it's an issue with a problem file and lots of students are using that problem file. In the latter case, it would be good to search for such things in the error log and go fix them. Is it possible to log whether a message comes from a problem rendered by the editor versus one rendered as part of regular use?

**Comment by drgrice1** [comment]
Some more warning messages will come through after https://github.com/openwebwork/pg/pull/709 is merged.  Although still not the ones you mention.

If the eval warnings come from "eval"ing a file (or really Safe "reval"ing) and are passed through the warning handlers then "eval ????" replaces the file name.  The problem is that some of the warnings are not passed through the warning handlers.  These may be able to be tracked down and perhaps fixed to replace "eval ????" with the file name.  However, there are also literal (r)eval's of strings that sometimes cause these warnings.  Those do not have a file name to associate with the "eval ????".

**Comment by pstaabp** [comment]
I've noticed that moving the macro files to PG causes more warnings to show up in the problem itself and not just the apache log.  I really don't understand why this happened.

**Comment by drgrice1** [comment]
That is a change in #709, not in the macro file move.  That is something we need to discuss.  I can move the warnings back to the apache error log, but really these warnings need to be fixed.  It was a bad idea from the beginning to hide all of the warnings that are occurring in problem files, macros, and even lib files.  That is why there are so many unexplained issues with PG.  PG really needs a lot of work.
